### ISSUE TITLE
[SQLAlchemy] Should the total be the count of query or length of resulted mappings?

### ISSUE THREAD

**Issue by rushilsrivastava** [uriyyo/fastapi-pagination#246]
In the [code](https://github.com/uriyyo/fastapi-pagination/blob/main/fastapi_pagination/ext/sqlalchemy.py#L29), the `total` parameter is defined as:

```Python
total = query.count()
```

Shouldn't this be `len(query.all())` instead? My understanding is that for someone who is consuming this API, the total returned instances isn't important because the API is only dealing with mapped object.

**Comment by uriyyo** [comment]
I am not sure about this one.

The purpose of `total` is to show total number of available entities.

**Comment by rushilsrivastava** [comment]
Yes. I'm the case of a query join, this would show more results Than the entities that would be paginated. When using length, it would show only the length of the paginated entities.

**Comment by uriyyo** [comment]
That's a bug but `len(query.all())` will lead to a performance issue cause it will load all available rows.

**Comment by rushilsrivastava** [comment]
Closing this because this is just a matter of optimizing queries.

**Comment by MattKravetz** [comment]
Hi, I recently ran into this issue with my API where an endpoint returning paginated responses would return an incorrect number of results if the query that contained an outer join that produced duplicate rows.  The duplication was handled well by SQLAlchemy/SQLModel to produce unique objects with nested relations, but the limit/offset pagination provided by this library returned the count based on the raw query result.  

As a workaround, I've updated my query to remove the outer join.  I'm not sure if/how to resolve this given the performance issues with a `len(query.all())` approach, but it may be useful to update documentation to call out this limitation (if it is not already mentioned). 

One possible solution could be to adapt the row-number pagination approach (https://github.com/uriyyo/fastapi-pagination/pull/370/files#diff-810070b01cb7d59769558a64771bd40c25378840b359ab97ccfd86db49094d93R21-R27) to partition on the primary key of the top-level entity.  I'm not sure if there's a clean way to do this without requiring `paginate` callers to pass additional context on the query (like the `pk` of the primary entity), and without limiting the types of queries that can be processed by this library. This would also have the downside of only solving for the `row-number` pagination approach.  

In any case, thanks for your work on this library!
